PhysicsTools/PatAlgos/plugins/PATTracksToPackedCandidates.cc

```diff
@@ -0,0 +1,210 @@
+                          const reco::VertexRef& pvSlimmed,
+                          const reco::VertexRefProd& pvSlimmedColl,
+                          bool passPixelTrackSel) const;
+
+private:
+  void produce(edm::StreamID, edm::Event&, const edm::EventSetup&) const override;
+
+  // ----------member data ---------------------------
+  const edm::EDGetTokenT<reco::TrackCollection> srcTracks_;
+  const edm::EDGetTokenT<reco::VertexCollection> srcPrimaryVertices_;
+  const edm::EDGetTokenT<reco::BeamSpot> srcOfflineBeamSpot_;
+  const double dzSigCut_;
+  const double dxySigCut_;
+  const double dzSigHP_;
+  const double dxySigHP_;
+  const double ptMax_;
+  const double ptMin_;
+  const bool resetHP_;
+  const int covarianceVersion_;
+  const int covarianceSchema_;
+};
+
+//
+// constants, enums and typedefs
+//
+
+//
+// static data member definitions
+//
+
+//
+// constructors and destructor
+//
+PATTracksToPackedCandidates::PATTracksToPackedCandidates(const edm::ParameterSet& iConfig)
+    : srcTracks_(consumes<reco::TrackCollection>(iConfig.getParameter<edm::InputTag>("srcTracks"))),
+      srcPrimaryVertices_(consumes<reco::VertexCollection>(iConfig.getParameter<edm::InputTag>("srcPrimaryVertices"))),
+      srcOfflineBeamSpot_(consumes<reco::BeamSpot>(iConfig.getParameter<edm::InputTag>("srcOfflineBeamSpot"))),
+      dzSigCut_(iConfig.getParameter<double>("dzSigCut")),
+      dxySigCut_(iConfig.getParameter<double>("dxySigCut")),
+      dzSigHP_(iConfig.getParameter<double>("dzSigHP")),
+      dxySigHP_(iConfig.getParameter<double>("dxySigHP")),
+      ptMax_(iConfig.getParameter<double>("ptMax")),
+      ptMin_(iConfig.getParameter<double>("ptMin")),
+      resetHP_(iConfig.getParameter<bool>("resetHP")),
+      covarianceVersion_(iConfig.getParameter<int>("covarianceVersion")),
+      covarianceSchema_(iConfig.getParameter<int>("covarianceSchema")) {
+  produces<std::vector<pat::PackedCandidate>>();
+  produces<edm::Association<pat::PackedCandidateCollection>>();
+}
+
+//
+// member functions
+//
+
+// ------------ method called to produce the data  ------------
+void PATTracksToPackedCandidates::produce(edm::StreamID, edm::Event& iEvent, const edm::EventSetup& iSetup) const {
+  using namespace edm;
+  using namespace std;
+
+  //track collection
+  auto tracks = iEvent.getHandle(srcTracks_);
+
+  auto outPtrTrksAsCands = std::make_unique<std::vector<pat::PackedCandidate>>();
+
+  //vtx collection
+  auto pvs = iEvent.getHandle(srcPrimaryVertices_);
+  reco::VertexRef pv(pvs.id());
+  reco::VertexRefProd pvRefProd(pvs);
+
+  //best vertex
+  double bestvzError;
+  math::XYZPoint bestvtx;
+  math::Error<3>::type vtx_cov;
+  if (!pvs->empty()) {
+    pv = reco::VertexRef(pvs, 0);
+    const reco::Vertex& vtx = (*pvs)[0];
+    bestvzError = vtx.zError();
+    bestvtx = vtx.position();
+    vtx_cov = vtx.covariance();
+  } else {
+    auto offlineBS = iEvent.getHandle(srcOfflineBeamSpot_);
+    reco::BeamSpot bs = *offlineBS;
```

Comment:
```suggestion
    const auto& bs = iEvent.get(srcOfflineBeamSpot_);
```

Reply:
Thanks for the comment. I included this in this commit: https://github.com/cms-sw/cmssw/pull/30313/commits/eaafc669cfe93d96463ff1aefd7dba7fb62907d9